File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.py
import sys
input = sys.stdin.readline
from collections import deque
n, m, k = map(int,input().split()) 
arr = []
graph = [[ ] for _ in range(n+1)]
q = deque()
visited = [0] * (n+1)
visited[k] = 1
cnt = 1
q.append(-k)
for i in range(m):
  a, b = map(int,input().split())
  graph[a].append(b)
  graph[b].append(a)
for i in graph:
  i.sort(reverse=True)
while q:
  x = -q.popleft()
  for i in graph[x]:
    if not visited[i]:
      q.append(-i)
      cnt += 1
      visited[i] = cnt
for i in visited[1:]:
  print(i)

File: 백준/Silver/24445. 알고리즘 수업 － 너비 우선 탐색 2/알고리즘 수업 － 너비 우선 탐색 2.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.ArrayDeque;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Deque;
import java.util.StringTokenizer;
public class Main {
	public static void main(String[] args) throws Exception{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        int k = Integer.parseInt(st.nextToken());
        Deque<Integer> q = new ArrayDeque<>();
        int[] visited = new int[n+1];
        q.add(k);
        visited[k] = 1;
        ArrayList<ArrayList<Integer>> graph = new ArrayList<>();
        for(int i = 0; i <= n; i++) {
            graph.add(new ArrayList<>());
        }
        for(int i = 0; i < m; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        int cnt = 1;
        // bfs
        while(!q.isEmpty()) {
            int x = q.poll();
            Collections.sort(graph.get(x), Collections.reverseOrder()); // 내림차순
            for (int num : graph.get(x)) {
                if (visited[num] == 0) { 
                    cnt+=1;
                    visited[num] = cnt;
                    q.add(num);
                }
            }
        }
        for(int i = 1; i <= n; i++) {
            System.out.println(visited[i]);
        }
    }
}